### PR TITLE
Fix: Handle Duplicate File Names in IOS Picker

### LINKED ISSUE #66: bug: File Picker Sending Duplicate Images when Selecting Downloaded Images from IOS
### Plugin(s)

- [ ] Android Foreground Service
- [ ] Android Battery Optimization
- [ ] App Update
- [ ] Background Task
- [ ] Badge
- [ ] Cloudinary
- [ ] Datetime Picker
- [ ] File Opener
- [X] File Picker
- [ ] Managed Configurations
- [ ] NFC
- [ ] Photo Editor
- [ ] Screen Orientation

### Did you test the latest version?

- [X] I use the latest version

### Platform(s)

- [ ] Android
- [X] iOS
- [ ] Web

### Current behavior

I have a few GoPro images that I downloaded on IOS. When I select the files from IOS Photos despite being different files the file-picker plugin sends me 2 duplicate images. Ive attached screenshots of the image / info below. When these are selected I get the following from the File Picker plugin.

Capacitor Logs:

```
⚡️  TO JS {"files":[{"name":"GPTempDownload.heic","size":1146838,"height":3286,"mimeType":"image\/heic","width":4288,"modifiedAt":1690988208860,"path":"file:\/\/\/var\/mobile\/Containers\/Data\/Application\/F13934F8-6B20-4F64-98CA-0F5E02FA221B\/Library\/Caches\/GPTe
```

Received in the JS Layer:

```
[
  {
    "name": "GPTempDownload.heic",
    "size": 1146838,
    "height": 3286,
    "mimeType": "image/heic",
    "width": 4288,
    "modifiedAt": 1690988208860,
    "path": "file:///var/mobile/Containers/Data/Application/F13934F8-6B20-4F64-98CA-0F5E02FA221B/Library/Caches/GPTempDownload.heic"
  },
  {
    "name": "GPTempDownload.heic",
    "size": 1146838,
    "height": 3286,
    "mimeType": "image/heic",
    "width": 4288,
    "modifiedAt": 1690988208860,
    "path": "file:///var/mobile/Containers/Data/Application/F13934F8-6B20-4F64-98CA-0F5E02FA221B/Library/Caches/GPTempDownload.heic"
  }
]
```

IMG_1845
IMG_1846

### Expected behavior

FilePIcker should not send duplicate images given the the files and different. My guess is it has something to do with the name being the same `GPTempDownload.heic` which appears to be the given Icloud name and not the actual name of the file. 

If needed I can provide the raw files.

### Reproduction

https://github.com/capawesome-team/.capacitor-app

### Steps to reproduce

Open FilePicker with GoPro Images or Images downloaded from Icloud on IOS.
Select the files
Observe that data is duplicated despite files being unique.

### Other information

It's hard to provide a reproduction because it has more to do with the files being selected. Again if needed I can provide the files directly to a member of the team. I think uploading them here would change the metadata that IOS provides.

Edited:

After doing a bit more research it turns out the files all have the same name GPTempDownload.heic and the caption is the unique value per the screenshots above. It seems like having duplicate names is throwing the logic off.

### Capacitor doctor

```
Latest Dependencies:

  @capacitor/cli: 5.2.2
  @capacitor/core: 5.2.2
  @capacitor/android: 5.2.2
  @capacitor/ios: 5.2.2

Installed Dependencies:

  @capacitor/cli: 5.0.5
  @capacitor/core: 5.2.2
  @capacitor/android: 5.2.2
  @capacitor/ios: 5.2.2
  ```

### Before submitting

- [X] I understand that incomplete issues (e.g. without reproduction) are closed.

### PR DESCRIPTION
**Pull request by danbunkr**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] The changes have been tested successfully.
- [ ] A changeset has been created (`npm run changeset`).

Close #66

### CODE REVIEW

**Comment by robingenz** [comment]
@danbunkr Thank you! I'll take a look at the PR after I get back from my vacation.

**Comment by danbunkr** [comment]
@robingenz - Just wanted to bump this and see what the consensus was for the PR

**Review by robingenz — CHANGES_REQUESTED** [review]
@danbunkr Thank you for your contribution!
TBH i don't like the approach that much because it's very hacky. ^^

My idea would be: Instead of saving all selected files in the same temporary folder, we create one folder per selected file with a UUID as folder name using the `saveTemporaryFile(...)` method. Then same file names won't cause problems anymore and we don't need to change the `getNameFromUrl(...)`.

**Comment by danbunkr** [comment]
Hey @robingenz. This is fair feedback on it being a a bit hacky. So I went and tried to implement your suggestion which I thought was a better idea of putting it in a unique folder with a uuid. However based on the logs and details im gonna show below it appears that the Capacitor.convertFileSrc() has issues with this approach. At this point I am kinda out of ideas but still committed to helping in anyway I can just let me know what you want to do from here.

Updated Swift Code:

```
    private func saveTemporaryFile(_ sourceUrl: URL) throws -> URL {
        let uniqueFolderName = UUID().uuidString
        var directory = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(uniqueFolderName)
        
        if let cachesDirectory = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first {
            directory = cachesDirectory.appendingPathComponent(uniqueFolderName)
        }
        // Create the unique folder if it doesn't exist
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true, attributes: nil)

        let targetUrl = directory.appendingPathComponent(sourceUrl.lastPathComponent)
        do {
            try deleteFile(targetUrl)
        }
        try FileManager.default.copyItem(at: sourceUrl, to: targetUrl)
        return targetUrl
    }

```

Javascript Code:

```
         const filePath =
              file.mimeType === FILE_CONTENT_TYPES.HEIC
                ? await FilePicker.convertHeicToJpeg({ path: file.path })
                : file;

            const mimeType =
              file.mimeType === FILE_CONTENT_TYPES.HEIC
                ? FILE_CONTENT_TYPES.JPEG
                : file.mimeType;

            const nativeSource = Capacitor.convertFileSrc(filePath.path);

            console.log('RESULT**', file?.path, nativeSource);
```


Result of Log:

```
⚡️  [log] - RESULT** file:///var/mobile/Containers/Data/Application/248AC400-6D44-4512-A70B-9C460CBD9463/Library/Caches/8D226F02-7AB3-4F67-B023-60298E6CCF08/GPTempDownload.heic capacitor://localhost/_capacitor_file_/private/var/mobile/Containers/Data/Application/248AC400-6D44-4512-A70B-9C460CBD9463/tmp/GPTempDownload.jpeg
⚡️  [log] - RESULT** file:///var/mobile/Containers/Data/Application/248AC400-6D44-4512-A70B-9C460CBD9463/Library/Caches/BFF53612-B016-4B3E-8009-2ABBAF92CBED/GPTempDownload.heic capacitor://localhost/_capacitor_file_/private/var/mobile/Containers/Data/Application/248AC400-6D44-4512-A70B-9C460CBD9463/tmp/GPTempDownload.jpeg
⚡️  [log] - RESULT** file:///var/mobile/Containers/Data/Application/248AC400-6D44-4512-A70B-9C460CBD9463/Library/Caches/D39B5C87-867F-43BE-BB58-B176A2BCBB2E/GPTempDownload.heic capacitor://localhost/_capacitor_file_/private/var/mobile/Containers/Data/Application/248AC400-6D44-4512-A70B-9C460CBD9463/tmp/GPTempDownload.jpeg

```

**Comment by danbunkr** [comment]
Ok @robingenz - I took one more shot at it by editing my code from yesterday and using your suggestion and got it working. The PR has been updated but interested to know your thoughts

File: packages/file-picker/ios/Plugin/FilePicker.swift [hunk 1]
```diff
@@ -213,11 +213,14 @@ import MobileCoreServices
...
-        var directory = URL(fileURLWithPath: NSTemporaryDirectory())
+        let uniqueFolderName = UUID().uuidString
+        var directory = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(uniqueFolderName)
         if let cachesDirectory = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first {
-            directory = cachesDirectory
+            directory = cachesDirectory.appendingPathComponent(uniqueFolderName)
         }
-        let targetUrl = directory.appendingPathComponent(sourceUrl.lastPathComponent)
+        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true, attributes: nil)
+        let uniqueFileName = "\(directory.lastPathComponent)_\(sourceUrl.lastPathComponent)"
```
Comment (robingenz) [inline packages/file-picker/ios/Plugin/FilePicker.swift:222]: Why is this line necessary? With this we change the original file name.

Comment (AdlerJS) [reply packages/file-picker/ios/Plugin/FilePicker.swift:222]: @robingenz - Without creating a unique fileName I was noticing that the Capacitor.convertFileSrc method would still point all the files to the same instance of a file resulting in duplicates. You can see more info on the comment on the main thread that shows both the code I tried to implement at the Swift / JS layer and the result that was being output.

For this way we change the file name to make it unique at a temporary file level but the getNameFromUrl still works because it just grabs the lastPathComponent and passes the correct value to the JS layer

Comment (robingenz) [reply packages/file-picker/ios/Plugin/FilePicker.swift:222]: You mean we need a unique file path? 
And we already have a unique folder name, so this should work without renaming the file itself.

Comment (robingenz) [reply packages/file-picker/ios/Plugin/FilePicker.swift:222]: Can you give it a try?

Comment (AdlerJS) [reply packages/file-picker/ios/Plugin/FilePicker.swift:222]: @robingenz  - Changed the function to:

```
    private func saveTemporaryFile(_ sourceUrl: URL) throws -> URL {
         let uniqueFolderName = UUID().uuidString
         var directory = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(uniqueFolderName)
         if let cachesDirectory = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first {
             directory = cachesDirectory.appendingPathComponent(uniqueFolderName)
         }
         try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true, attributes: nil)
         let targetUrl = directory.appendingPathComponent(sourceUrl.lastPathComponent)
         do {
             try deleteFile(targetUrl)
         }
         try FileManager.default.copyItem(at: sourceUrl, to: targetUrl)
         return targetUrl
     }
 ```
     
  And then logged out what the nativeSrc was returning after picking 3 files with the same name:
  
  ```
  ⚡️  [log] - NATIVE SRC capacitor://localhost/_capacitor_file_/private/var/mobile/Containers/Data/Application/3E00ADEE-62F0-416D-99CB-F337BB359275/tmp/GPTempDownload.jpeg
⚡️  [log] - NATIVE SRC capacitor://localhost/_capacitor_file_/private/var/mobile/Containers/Data/Application/3E00ADEE-62F0-416D-99CB-F337BB359275/tmp/GPTempDownload.jpeg
⚡️  [log] - NATIVE SRC capacitor://localhost/_capacitor_file_/private/var/mobile/Containers/Data/Application/3E00ADEE-62F0-416D-99CB-F337BB359275/tmp/GPTempDownload.jpeg

```

They are duplicating the same as they are placed in the same tmp folder with the same name

Comment (AdlerJS) [reply packages/file-picker/ios/Plugin/FilePicker.swift:222]: I also just validated that the `targetUrl` is unique as well by logging it out at the native level

```
HELP file:///var/mobile/Containers/Data/Application/1DEEE4F5-97D8-4327-92F7-878B8A0BA8FF/Library/Caches/941A951B-E07D-4E1A-B1CB-BB3318FBBD76/GPTempDownload.heic
HELP file:///var/mobile/Containers/Data/Application/1DEEE4F5-97D8-4327-92F7-878B8A0BA8FF/Library/Caches/5ADCE43B-0236-4D35-847A-3A59273F6824/GPTempDownload.heic
HELP file:///var/mobile/Containers/Data/Application/1DEEE4F5-97D8-4327-92F7-878B8A0BA8FF/Library/Caches/CF10D8D2-8DE4-4C60-AABC-8BE154DEBB07/GPTempDownload.heic
```

Comment (robingenz) [reply packages/file-picker/ios/Plugin/FilePicker.swift:222]: > I also just validated that the `targetUrl` is unique as well by logging it out at the native level
> 
> ```
> HELP file:///var/mobile/Containers/Data/Application/1DEEE4F5-97D8-4327-92F7-878B8A0BA8FF/Library/Caches/941A951B-E07D-4E1A-B1CB-BB3318FBBD76/GPTempDownload.heic
> HELP file:///var/mobile/Containers/Data/Application/1DEEE4F5-97D8-4327-92F7-878B8A0BA8FF/Library/Caches/5ADCE43B-0236-4D35-847A-3A59273F6824/GPTempDownload.heic
> HELP file:///var/mobile/Containers/Data/Application/1DEEE4F5-97D8-4327-92F7-878B8A0BA8FF/Library/Caches/CF10D8D2-8DE4-4C60-AABC-8BE154DEBB07/GPTempDownload.heic
> ```

However, with a unique `targetUrl` it should work or not? In your issue the problem was that the `path` was the same for both files. We have fixed this now with the unique folder names. The files are no longer overwritten. The thing with `nativeSrc` could be a different problem.
Or am I missing something?

Comment (AdlerJS) [reply packages/file-picker/ios/Plugin/FilePicker.swift:222]: @robingenz - I think you are right but we're needing some additional logic that I am not sure where to add. When logging out the filepath at the JS layer and the native source I'm getting:

```
⚡️ [log] - FILE PATH: {"path":"file:///private/var/mobile/Containers/Data/Application/09E61D33-6E45-41D1-80CD-5AD5FA10CE5F/tmp/GPTempDownload.jpeg"}
   capacitor://localhost/_capacitor_file_/private/var/mobile/Containers/Data/Application/09E61D33-6E45-41D1-80CD-5AD5FA10CE5F/tmp/GPTempDownload.jpeg

⚡️ [log] - FILE PATH: {"path":"file:///private/var/mobile/Containers/Data/Application/09E61D33-6E45-41D1-80CD-5AD5FA10CE5F/tmp/GPTempDownload.jpeg"}
   capacitor://localhost/_capacitor_file_/private/var/mobile/Containers/Data/Application/09E61D33-6E45-41D1-80CD-5AD5FA10CE5F/tmp/GPTempDownload.jpeg

⚡️ [log] - FILE PATH: {"path":"file:///private/var/mobile/Containers/Data/Application/09E61D33-6E45-41D1-80CD-5AD5FA10CE5F/tmp/GPTempDownload.jpeg"}
   capacitor://localhost/_capacitor_file_/private/var/mobile/Containers/Data/Application/09E61D33-6E45-41D1-80CD-5AD5FA10CE5F/tmp/GPTempDownload.jpeg

```

Which tells me that despite the temporaryFile being unique there is some logic specifically around building the path that is not correct. What's happening is we are now getting 3 URLs but they are all pointing to the same file and it's still duplicating on the frontend
